test/ccall.jl

```diff
@@ -1509,3 +1509,111 @@ let
     @test isa(ptr, Ptr{Cvoid})
     @test arr[1] == '0'
 end
+
+# @ccall macro
+using Base: @ccall, ccall_macro_parse, ccall_macro_lower
+@testset "test basic ccall_macro_parse functionality" begin
+    callexpr = :(
+        libc.printf("%s = %d\n"::Cstring ; name::Cstring, value::Cint)::Cvoid
+    )
+    @test ccall_macro_parse(callexpr) == (
+        :((:printf, libc)),               # function
+        :Cvoid,                           # returntype
+        Any[:Cstring, :Cstring, :Cint],   # argument types
+        Any["%s = %d\n", :name, :value],  # argument symbols
+        1                                 # number of required arguments (for varargs)
+    )
+end
+
+@testset "ensure the base-case of @ccall works, including library name and pointer interpolation" begin
+    call = ccall_macro_lower(:ccall, ccall_macro_parse( :( libstring.func(
+        str::Cstring,
+        num1::Cint,
+        num2::Cint
+    )::Cstring))...)
+    @test call == Base.remove_linenums!(
+        quote
+        arg1root = Base.cconvert($(Expr(:escape, :Cstring)), $(Expr(:escape, :str)))
+        arg1 = Base.unsafe_convert($(Expr(:escape, :Cstring)), arg1root)
+        arg2root = Base.cconvert($(Expr(:escape, :Cint)), $(Expr(:escape, :num1)))
+        arg2 = Base.unsafe_convert($(Expr(:escape, :Cint)), arg2root)
+        arg3root = Base.cconvert($(Expr(:escape, :Cint)), $(Expr(:escape, :num2)))
+        arg3 = Base.unsafe_convert($(Expr(:escape, :Cint)), arg3root)
+        $(Expr(:foreigncall,
+               :($(Expr(:escape, :((:func, libstring))))),
+               :($(Expr(:escape, :Cstring))),
+               :($(Expr(:escape, :(($(Expr(:core, :svec)))(Cstring, Cint, Cint))))),
+               0,
+               :(:ccall),
+               :arg1, :arg2, :arg3, :arg1root, :arg2root, :arg3root))
+        end)
+
+    # pointer interpolation
+    call = ccall_macro_lower(:ccall, ccall_macro_parse(:( $(Expr(:$, :fptr))("bar"::Cstring)::Cvoid ))...)
+    @test Base.remove_linenums!(call) == Base.remove_linenums!(
+    quote
+        func = $(Expr(:escape, :fptr))
+        begin
+            if !(func isa Ptr{Nothing})
+                name = :func
+                throw(ArgumentError("interpolated function `$(name)` was not a Ptr{Cvoid}, but $(typeof(func))"))
+            end
+        end
+        arg1root = Base.cconvert($(Expr(:escape, :Cstring)), $(Expr(:escape, "bar")))
+        arg1 = Base.unsafe_convert($(Expr(:escape, :Cstring)), arg1root)
+        $(Expr(:foreigncall, :func, :($(Expr(:escape, :Cvoid))), :($(Expr(:escape, :(($(Expr(:core, :svec)))(Cstring))))), 0, :(:ccall), :arg1, :arg1root))
+    end)
+
+end
+
+@testset "check error paths" begin
+    # missing return type
+    @test_throws ArgumentError ccall_macro_parse(:( foo(4.0::Cdouble )))
```

Comment:
If you feel like it you could make these tests more precise using a value rather than a type for the exception; `@test_throws ArgumentError("some error message") ...`

Reply:
resolved in https://github.com/JuliaLang/julia/pull/32748/commits/a651fbe95a93741300bf6d71c89faf0fc62692f9